google/cloud/internal/grpc_opentelemetry.h

```diff
@@ -0,0 +1,49 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_GRPC_OPENTELEMETRY_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_INTERNAL_GRPC_OPENTELEMETRY_H
+
+#include "google/cloud/version.h"
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include <opentelemetry/nostd/shared_ptr.h>
+#include <opentelemetry/trace/span.h>
+#endif  // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+
+/**
+ * Make a span, setting attributes related to gRPC.
+ *
+ * @see
+ * https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/rpc/
+ * for the semantic conventions used for span names and attributes.
+ */
+opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> MakeSpanGrpc(
+    opentelemetry::nostd::string_view service,
+    opentelemetry::nostd::string_view method);
```

Comment:
Here is a scary though: should we capture the target resources: all that stuff the metadata decorator puts in headers.

Reply:
Probably.

It is scarier than I originally thought. My vague plan was to capture this information outside of the tracing decorators. (the `grpc::ClientContext` does not give us access to the metadata we have set, so we could not do it at the end of the call). I would use something like: `AddSpanAttribute(...)` in the MD decorator: https://github.com/dbolduc/google-cloud-cpp/blob/9e9691476a167626486dba50efab0a7eafef2334/google/cloud/internal/opentelemetry.cc#L98-L106

But it is much better if the resource is known **when we create the span**. That way, the customers can set up a [sampler](https://opentelemetry.io/docs/reference/specification/trace/sdk/#sampler) to filter on a resource name.  (sampling happens when spans are created).

<hr />

Barely-related Aside: The [spec](https://opentelemetry.io/docs/reference/specification/trace/semantic_conventions/rpc/#grpc-request-and-response-metadata) says: "Instrumentations SHOULD require an explicit configuration of which metadata values are to be captured. Including all request/response metadata values can be a security risk - explicit configuration helps avoid leaking sensitive information".

Reply:
Hmm... Ack.  I think those should be used by the gRPC team.  Maybe we should have one in a different namespace, maybe: `rpc.gcloud-cpp.resource`

In any case, something to add later. Feel free to commit as-is.

Reply:
> `grpc::ClientContext` does not give us access to the metadata we have set

```
namespace grpc {
namespace testing {

class ClientContextTestPeer {
 public:
  explicit ClientContextTestPeer(ClientContext* const ctx) : ctx_(ctx) {}

  std::multimap<std::string, std::string> GetSendInitialMetadata() const {
    return ctx_->send_initial_metadata_;
  } 
  
 private:
  ClientContext* const ctx_;  
};

}  // namespace testing
}  // namespace grpc
```
:-)